sdks/java/io/hadoop-format/src/main/java/org/apache/beam/sdk/io/hadoop/format/HadoopFormatIO.java

```diff
@@ -101,94 +133,290 @@
+     * storing of all {@link KV} pairs.
+     *
+     * @see #DEFAULT_CONFIG_COMBINE_FN
+     * @return Combine function
+     */
+    default CombineFnBase.GlobalCombineFn<Configuration, ?, Configuration> getConfigCombineFn() {
+      return DEFAULT_CONFIG_COMBINE_FN;
+    }
+  }
+
+  /**
+   * Default implementation of Configuration transform. It requires only particular {@link
+   * PTransform} to be specified.
+   *
+   * @param <KeyT> Key type which should be written
+   * @param <ValueT> Value type which should be written
+   */
+  private static class DefaultConfigurationTransform<KeyT, ValueT>
+      implements IConfigurationTransform<KeyT, ValueT> {
+
+    private PTransform<PCollection<? extends KV<KeyT, ValueT>>, PCollection<Configuration>>
+        configTransform;
+
+    public DefaultConfigurationTransform(
+        PTransform<PCollection<? extends KV<KeyT, ValueT>>, PCollection<Configuration>>
+            configTransform) {
+      this.configTransform = configTransform;
+    }
 
-  public static final String OUTPUTFORMAT_CLASS = "mapreduce.job.outputformat.class";
-  public static final String OUTPUTFORMAT_KEY_CLASS = "mapreduce.job.outputformat.key.class";
-  public static final String OUTPUTFORMAT_VALUE_CLASS = "mapreduce.job.outputformat.value.class";
+    @Override
+    public PTransform<PCollection<? extends KV<KeyT, ValueT>>, PCollection<Configuration>>
+        getConfigTransform() {
+      return configTransform;
+    }
+  }
 
   /**
-   * Creates an uninitialized {@link HadoopFormatIO.Write}. Before use, the {@code Write} must be
-   * initialized with a HadoopFormatIO.Write#withConfiguration(HadoopConfiguration) that specifies
-   * the sink.
+   * Generates tasks for output pairs and groups them by this key.
+   *
+   * <p>This transformation is used when is configured write with partitioning.
+   *
+   * @param <KeyT> type of key
+   * @param <ValueT> type of value
    */
-  public static <K, V> Write<K, V> write() {
-    return new AutoValue_HadoopFormatIO_Write.Builder<K, V>().build();
+  private static class GroupDataByPartition<KeyT, ValueT>
+      extends PTransform<
+          PCollection<KV<KeyT, ValueT>>, PCollection<KV<Integer, KV<KeyT, ValueT>>>> {
+
+    private PCollectionView<Configuration> configView;
+
+    private GroupDataByPartition(PCollectionView<Configuration> configView) {
+      this.configView = configView;
+    }
+
+    @Override
+    public PCollection<KV<Integer, KV<KeyT, ValueT>>> expand(PCollection<KV<KeyT, ValueT>> input) {
+      return input
+          .apply(
+              "AssignTask",
+              ParDo.of(new AssignTaskFn<KeyT, ValueT>(configView)).withSideInputs(configView))
+          .setTypeDescriptor(
+              TypeDescriptors.kvs(TypeDescriptors.integers(), input.getTypeDescriptor()))
+          .apply("GroupByTaskId", GroupByKey.create())
+          .apply(
+              "FlattenGroupedTasks",
+              FlatMapElements
```

Comment:
I believe this transformation is wrong as it requires whole iterable to fit in memory (it needs to be collected to list). I guess the most straightforward approach to address the issue would be flatten using ParDo:

```java

  private static class FlattenGroupedTasks<KeyT, ValueT>
      extends DoFn<KV<Integer, Iterable<KV<KeyT, ValueT>>>, KV<Integer, KV<KeyT, ValueT>>> {

    @ProcessElement
    public void processElement(
        @Element KV<Integer, Iterable<KV<KeyT, ValueT>>> input,
        OutputReceiver<KV<Integer, KV<KeyT, ValueT>> outputReceiver) {
      final Integer key = input.getKey();
      for (KV<KeyT, ValueT> element : Objects.requireNonNull(input.getValue(), "Iterable can not be null.")) {
        outputReceiver.output(KV.of(key, element));
      }
    }

  }
```

Reply:
I got it, thanks